superset/errors.py

```diff
@@ -46,6 +47,23 @@ class SupersetErrorType(str, Enum):
     DATASOURCE_SECURITY_ACCESS_ERROR = "DATASOURCE_SECURITY_ACCESS_ERROR"
     MISSING_OWNERSHIP_ERROR = "MISSING_OWNERSHIP_ERROR"
 
+    # Other errors
+    BACKEND_TIMEOUT_ERROR = "BACKEND_TIMEOUT_ERROR"
+
+
+ERROR_TYPES_TO_ERROR_CODES_MAPPING = {
+    SupersetErrorType.BACKEND_TIMEOUT_ERROR: [
+        {
+            "code": 1000,
+            "message": _("Error 1000 - The datasource is too large to query."),
+        },
+        {
+            "code": 1001,
+            "message": _("Error 1001 - The database is under an unusual load."),
+        },
+    ]
+}
```

Comment:
Hmm, just realized a bigger question: are there cases that the backend will be able to tell which is the actual reason? If not, why can't they be the same error code?

```
## Error 1000 - Timeout communicating with datasource.

Your datasource may be too large to query or was under an unusual load.
```

Reply:
It's also possible to remove such mapping altogether:

```python
class SupersetError(IntEnum):
  BACKEND_TIMEOUT = 1001

class SupersetException(Exception):
    code = None
    
    def __init__(self, message=None, description=None):
        if not self.code:
           raise RuntimeError('Must not raise SupersetException directly')
        error_name = SupersetError(self.code).name
        self.message = message or _(f'error.{error_name}.message')
        self.description = description or _(f'error.{error_name}.message')

class SupersetTimeoutException(SupersetException):
    code = SupersetErrorEnum.BACKEND_TIMEOUT
```

Then add the detailed message and explanation in i18n.

Reply:
I think this again points to the differentiation between the actual error and potential root causes. In this case, `Error 1001` might be the root cause for many different errors (a timeout error, a generic DB error, a 503/504 from the db engine, etc.). That was the reason for splitting them into two different structures. However, I agree it's getting into a zone of too many different abstraction layers with both the error_type and the error_codes...

@steejay, this is something where your thoughts would be helpful: e.g. what relationship do you see between error types and the in product user facing error codes?

Reply:
The goal of the error code is to suggest a path to a solution by highlighting the potential sources. Ideally we'd like to design a message that can pinpoint the source but this can't always be achieved so instead we try to provide as much context as possible to let the technical user troubleshoot. But youre right, the same error source _can_ lead to different errors.

Reply:
I've expanded my thoughts a little and here's a more detailed example: https://gist.github.com/ktmud/59ccbdf6c96570e1ffe41ee8e5ecd479

Let me know if anything was unclear.

Even if an error may represent different root causes, we can return a generic error code and potential causes in `extras` without adding another layer of abstraction.

Reply:
Do you not find it a bit inconsistent that the same object is used both for the error code and for possible causes? These seem like they should be different things, just as a symptom (a bad cough) is different from a root cause (like the flu or allergies). We want the docs to provide remedies for the flu or allergies, and not simply prescribe cough drops.

My struggle is how to manage this mapping in the cleanest possible way. To me it seems like error_type and error_codes need to be seperate objects, each with different limitations (every error only has 1 error type, while it could have multiple error codes since they represent potential root causes).

Perhaps this is a naming issue. If these were named error_causes, and we displayed `Cause 1000` and `Cause 1001` within the box, would this be clearer?

Reply:
Yup, I think this is a great analogy @etr2460 

The renaming suggestion is an interesting idea and worth considering. I wonder if renaming to "Issue" instead would make more sense? 

This may be triggered by:
Issue 1000 - The datasource is too large to query
Issue 1001 - The database is under an unusual load

The statement "The datasource is too large to query" implies an issue. Not necessarily a cause or source of anything until applied to the context of this specific alert message.

Reply:
Yes, I agree error codes and possible root causes shouldn't be mixed together as suggesting root causes and providing remedies are more of a documentation issue rather than something the program cares about.

I've removed the `description` field in my example code, but other parts of the code still feel like an improvement to me though. There really is no need to have two separate entities for exception catching and messaging.

As for the mapping between symptoms and root causes, maybe we can extend the `errors.json` file I suggested earlier and add a `causes` field?

```js
{
  "issues": {
    "DATASOURCE_TOO_LARGE": {
      "code": 1001,
      "title": "The database is too large to query",
      "description": "It's likely your datasource has grown too large to run the current query, and is timing out. You can resolve this by reducing the size of your datasource or by modifying your query to only process a subset of your data."
    },
    "DATASOURCE_OVERLOAD": {
      "code": 1002,
      // ....
    },
  },
  "causes": {
    "BACKEND_TIMEOUT_ERROR": ["DATASOURCE_TOO_LARGE", "DATASOURCE_OVERLOAD"],
  }
}
```

Reply:
I think we might need separate entities for exception catching and error messaging. One example is for async queries in sql lab. We would catch an exception from the query failing, but then only want to write a subset of attributes to the query table for returning to the client later. Then when the client polls for query status, we're returning an error payload within that request, but no exception is occuring. They seem like 2 different entities to me, and I'm not sure how much is improved by having a complex inheritance tree consisting of both native Python Exceptions and UI specific payload classes

Reply:
They are indeed two entities, one is runtime exception, another is the result payload used by the API. But it doesn't mean the payload cannot be generated by exception itself.

The improvement is that you don't have to manually generate the payload every time an exception occurs. You just store relevant information as top-level attributes on the Exception and some general error handlers will return a consistent payload whenever needed. The payload doesn't have to be UI specific. We can certainly return different payloads for storage and API by adding arguments to the `to_json` method, if that's what we want.

I just find it weird that you had to initialize two classes in order to raise an API-conforming exception. We already use class inheritance for Exceptions and used the pattern extensively for `superset.commands.exceptions`, so I'm not sure an additional mixin to serialize the exception is that much of complexity.

Reply:
@villebro @dpgaspar curious to hear your thoughts on this, too.

Reply:
hmm, i haven't seen the `superset.commands.exceptions` module yet, i'll look at that and see how other parts of the code handles errors. This has been a wild west of sorts on the backend side, and I've mainly been focused on making sure whatever happens on the backend, the frontend gets something consistent. If there's a more standard approach now (i think it was discussed in SIP-41) then i'm happy to follow it